pipeline_steps/common.groovy

```diff
@@ -1,71 +1,88 @@
-      CURL_CMD="curl --silent --show-error --retry 5"
-      OUTPUT_FILE="get-pip.py"
-      \${CURL_CMD} https://bootstrap.pypa.io/get-pip.py > \${OUTPUT_FILE} \
-        || \${CURL_CMD} https://raw.githubusercontent.com/pypa/get-pip/master/get-pip.py > \${OUTPUT_FILE}
-      GETPIP_OPTIONS="pip setuptools wheel --constraint rpc-gating/constraints.txt"
-      \${VENV_PYTHON} \${OUTPUT_FILE} \${GETPIP_OPTIONS} \
-        || \${VENV_PYTHON} \${OUTPUT_FILE} --isolated \${GETPIP_OPTIONS}
-    fi
+      # Install Pip
+      set +x; source .venv/bin/activate; set -x
+
+      # UG-613 change TMPDIR to directory with more space
+      export TMPDIR="/var/lib/jenkins/tmp"
+
+      # If the pip version we're using is not the same as the constraint then replace it
+      PIP_TARGET="\$(awk -F= '/^pip==/ {print \$3}' rpc-gating/constraints.txt)"
+      VENV_PYTHON=".venv/bin/python"
+      VENV_PIP=".venv/bin/pip"
+      if [[ "\$(\${VENV_PIP} --version)" != "pip \${PIP_TARGET}"* ]]; then
+        # Install a known version of pip, setuptools, and wheel in the venv
+        CURL_CMD="curl --silent --show-error --retry 5"
+        OUTPUT_FILE="get-pip.py"
+        \${CURL_CMD} https://bootstrap.pypa.io/get-pip.py > \${OUTPUT_FILE} \
+          || \${CURL_CMD} https://raw.githubusercontent.com/pypa/get-pip/master/get-pip.py > \${OUTPUT_FILE}
+        GETPIP_OPTIONS="pip setuptools wheel --constraint rpc-gating/constraints.txt"
+        \${VENV_PYTHON} \${OUTPUT_FILE} \${GETPIP_OPTIONS} \
+          || \${VENV_PYTHON} \${OUTPUT_FILE} --isolated \${GETPIP_OPTIONS}
+      fi
 
-    # Install rpc-gating requirements
-    PIP_OPTIONS="-c rpc-gating/constraints.txt -r rpc-gating/requirements.txt"
-    \${VENV_PIP} install \${PIP_OPTIONS} \
-      || \${VENV_PIP} install --isolated \${PIP_OPTIONS}
-  """
-}
+      # Install rpc-gating requirements
+      PIP_OPTIONS="-c rpc-gating/constraints.txt -r rpc-gating/requirements.txt"
+      \${VENV_PIP} install \${PIP_OPTIONS} \
+        || \${VENV_PIP} install --isolated \${PIP_OPTIONS}
 
-def install_ansible_roles(){
-  print "install_ansible_roles"
-  sh """#!/bin/bash -xe
-    cd ${env.WORKSPACE}
-    . .venv/bin/activate
-    mkdir -p rpc-gating/playbooks/roles
-    ansible-galaxy install -r rpc-gating/role_requirements.yml -p rpc-gating/playbooks/roles
+      # Install ansible roles
+      mkdir -p rpc-gating/playbooks/roles
+      ansible-galaxy install -r rpc-gating/role_requirements.yml -p rpc-gating/playbooks/roles
+    }
+
+    download_venv(){
+      curl -s "\${REPO_BASE}/rpcgatingvenv_\${SHA}.tbz" > venv.tbz
+      tar xjfp venv.tbz
+      # virtualenv --relocatable doesn't update the activate script FGJ.
+      sed -i.bak 's|VIRTUAL_ENV=.*\$|VIRTUAL_ENV=${env.WORKSPACE}/.venv|' .venv/bin/activate
+      # CentOS 6 can take a hike, its glibc isn't new enough for python 2.7.12
+      if which scl; then
+        cp /opt/rh/python27/root/usr/bin/python .venv/bin/python
```

Comment:
Isn't that what we are doing now already?
Let's try and see what happens? In any case we'll probably have to iterate there.